addons/sale_stock_margin/models/sale_order_line.py

```diff
@@ -1,26 +1,28 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models
+from odoo import api, models
 
 
 class SaleOrderLine(models.Model):
     _inherit = 'sale.order.line'
 
     @api.depends('move_ids', 'move_ids.stock_valuation_layer_ids', 'move_ids.picking_id.state')
     def _compute_purchase_price(self):
-        lines_without_moves = self.browse()
+        lines_to_pass = self.browse()
         for line in self:
+            # ignore lines without moves or lines for std cost products (no valuation action needed)
+            if (not line.has_valued_move_ids() or
+                line.product_id.with_company(line.company_id).categ_id.property_cost_method == 'standard'
+            ):
+                lines_to_pass += line
+                continue
```

Comment:
```suggestion
        line_ids_to_pass = set()
        for line in self:
            # ignore lines without moves or lines for std cost products (no valuation action needed)
            if (not line.has_valued_move_ids() or
                line.product_id.with_company(line.company_id).categ_id.property_cost_method == 'standard'
            ):
                line_ids_to_pass.add(line.id)
                continue
        # [...]
        lines_to_pass = self.env['sale.order.line'].browse(line_ids_to_pass)
        return super(SaleOrderLine, lines_to_pass)._compute_purchase_price()
```
When possible, it's always better to avoid aggregating records again and again. I don't remember the specifics, but it has some performance issues when dealing with a large number of records.